RobotControl/Robot.cpp

```diff
@@ -50,41 +61,115 @@ void Robot::update() {
   if (timeLeftMillis > 0) {
-    delay(timeLeftMillis);
+    //delay(timeLeftMillis);
   }
 }
 
 void Robot::updateSerial() {
-  if (Serial.available() > 0) {
-    String line = Serial.readString();
-    updateFromJson(line.c_str());
-    m_state.printJson(Serial);
-    Serial.println();
+  memset(m_payloadBytes, 0, PAYLOAD_LEN);
+  memset(m_serialBuffer, 0, SERIAL_BUFFER_LEN);
+
+  if (m_commsI2C.getBytes(m_serialBuffer, SERIAL_BUFFER_LEN, m_payloadBytes, PAYLOAD_LEN)) {
+    updatePayload(m_payloadBytes, PAYLOAD_LEN);
+    //Serial.println("Bytes parsed");
   }
+
+  setStatus();
+  setRobot();
+  
+  //m_payload.print();
+  //Serial.println();
 }
 
-void Robot::updateFromJson(const char *json) {
-  m_state.updateFromJson(json);
+void Robot::updatePayload(const uint8_t *data, const uint8_t len) {
+  bool success = m_payload.readMessage(data, len);
 
-  const char *status = m_state.root()["status"].asString();
-  const char *err = m_state.root()["err"].asString();
+  const uint8_t status = m_payload.getStatus();
+  const uint8_t err = m_payload.getError();
 
-  if (strlen(err) > 0) {
+  if (err > 0 || !success) {
     m_statusLEDs.setBlinkPattern(StatusLEDs::ERROR);
-    m_state.root()["status"] = STATUS_DISABLED;
+    m_payload.setStatus(STATUS_DISABLED);
   } else {
-    if (strcmp(status, STATUS_DISABLED) == 0) {
+    if (status == STATUS_DISABLED) {
       m_statusLEDs.setBlinkPattern(StatusLEDs::DISABLED);
-    } else if (strcmp(status, STATUS_ENABLED) == 0) {
+    } else if (status == STATUS_ENABLED) {
       m_statusLEDs.setBlinkPattern(StatusLEDs::ENABLED);
-    } else if (strcmp(status, STATUS_PRIMED) == 0) {
+    } else if (status == STATUS_PRIMED) {
       m_statusLEDs.setBlinkPattern(StatusLEDs::PRIMED);
     } else {
       m_statusLEDs.setBlinkPattern(StatusLEDs::OFF);
     }
   }
 }
 
+void Robot::setRobot() {
+  uint8_t status = m_payload.getStatus();
+
+  if(m_firing) {
+    if(millis() - TEMP_FIRE_TIME_MILLIS >= m_solenoidOpenMillis) {
+      digitalWrite(m_fireSolenoidPin, LOW);
+      m_firing = false;
+      Serial.print("Open for: ");
+      Serial.println(millis() - m_solenoidOpenMillis);
+    }
+  }
+
+  if (status != STATUS_ENABLED) {
+    Serial.write((uint8_t)0);
+    Serial.write((uint8_t)128);
+  } 
+
+  if (status != STATUS_FIRING && status != STATUS_ADJUSTING) {
+    digitalWrite(m_fireSolenoidPin, LOW);
+    m_firing = false;
+    m_isHoldingFire = false;
+  }
+
+  if (status == STATUS_ENABLED) {
+    Serial.write(m_payload.getControllerDriveLeft());
+    Serial.write(m_payload.getControllerDriveRight());
+  } 
+
+  if (status == STATUS_FIRING) {
+    if(!m_isHoldingFire) {
+      digitalWrite(m_fireSolenoidPin, HIGH);
+      m_solenoidOpenMillis = millis();
+      status = STATUS_ADJUSTING;
+      Serial.println("Firing");
+      m_firing = true;
+      m_isHoldingFire = true;
+    }
+  }
+  m_payload.setStatus(status);
+}
+
+void Robot::setStatus() {
```

Comment:
Same here, should be `updateStatus` if using internal state, `setStatus` should take a payload or a status